src/services/suggestionDiagnostics.ts

```diff
@@ -120,17 +130,29 @@ namespace ts {
         return isBinaryExpression(commonJsModuleIndicator) ? commonJsModuleIndicator.left : commonJsModuleIndicator;
     }
 
-    function containsAsync(arr: NodeArray<Modifier>): boolean{
-        for(let modifier of arr){
-            if(modifier.kind === SyntaxKind.AsyncKeyword){
+    function isPromiseType(T: Type): boolean {
+        return T.flags === TypeFlags.Object && T.symbol.name === "Promise";
+    }
+
+    function hasCallback(stmt: Node): boolean {
+        if (stmt.kind === SyntaxKind.CallExpression && (<CallExpression>stmt).expression.kind === SyntaxKind.PropertyAccessExpression &&
+            ((<PropertyAccessExpression>(<CallExpression>stmt).expression).name.text === "then" || (<PropertyAccessExpression>(<CallExpression>stmt).expression).name.text === "catch")) {
                 return true;
-            }
         }
 
-        return false;
+        for (const child of stmt.getChildren().filter(node => isNode(node))) {
+            return hasCallback(child);
+        }
     }
 
-    function isPromiseType(T:Type):boolean{
-        return T.flags === TypeFlags.Object && T.symbol.name === "Promise";
+    function getReturnStmts(node: Node, retStmts: ReturnStatement[]) {
+        for (const child of node.getChildren().filter(node => isNode(node))) {

```

Comment:
no need to getChildren. this is a call that will reiterate the tokens in the tree, and since we do not care about tokens, only return statement nodes, we do not need to do that, and we do not need the `.filter` either.

also, you can use `forEachReturnStatement` instead.